### ISSUE TITLE
Proxy check scheduling seems unreliable through cluster restarts

### ISSUE THREAD

**Issue by echlebek** [sensu/sensu-go#4011]
## Expected Behavior
Proxy check scheduling should remain correct and available even when there are backend or agent restarts.

## Current Behavior
Backend and agent restarts tend to damage the scheduler. It can return to normal after a period of time, but it tends to take ~30 minutes.

## Possible Solution
Unknown

## Steps to Reproduce (for bugs)
Create a Sensu cluster and schedule many proxy checks. Cycle the backends and agends off and on. Observe the proxy check execution schedule.

**Comment by palourde** [comment]
Probably related to #4013

**Comment by mcbsd** [comment]
I added our observation into #4013

**Comment by portertech** [comment]
Hello everyone :wave: Thank you for your input on this issue. The Sensu Engineering team now has an understanding of why round-robin scheduling is inconsistent and potentially unreliable in OSS builds of Sensu Go. @echlebek has written up an excellent summary of our understanding and provides some advice:

Experience has shown that in some circumstances, sensu-go's scheduling of round robin and proxy round robin checks can be unreliable after cluster or agent restarts. Check scheduling can be delayed or incorrect. Detailed investigation has revealed that the use of the etcd "lease" mechanism for this type of scheduling can lead to this unreliable scheduling.

When sensu-go uses round-robin scheduling, it creates sets of agent entities called "rings" to execute checks on. These rings are composed of one or more leased keys in etcd. The intention is that unless the key is continually updated (i.e. on its keepalive schedule) the key will be removed from the ring, and the round robin scheduler will not consider it for execution. Thus, whenever an agent severs its connection to the server, its ring key is immediately removed. If the connection appears to be maintained, but the agent is unresponsive, then its key will be removed after the lease expires.

In a stable cluster, this method generally works. However, there are scenarios where this method breaks down and stops working.

1. etcd cluster is shut down, or cluster loses majority. In this circumstance, leases can be lost by etcd. Depending on the version of etcd, and depending on the length of the leased keys, the lease may be persisted to storage, or it may be held in memory. If it is held in memory, then the lease will be lost when the cluster restarts. When this happens, these stale ring keys can lead to check requests being sent into the void, with no corresponding agent to receive them, if the agents corresponding to those keys are no longer available.
2. Agents are mass-disconnected from a cluster that is on the brink of failure. If etcd is experiencing performance problems, and a large number of agents disconnect suddenly, then the cluster can fail to deal with the incoming load, and fail to remove all of the leased keys. This usually leads to a cluster restart as well, and then the above situation applies.
3. etcd is under heavy sustained load. When etcd is under intense, sustained load, then lease expirations are given a lower priority than serving GET and PUT requests. This can lead to the system expiring leases only after a very long time passes, sometimes 30-60 minutes. If a large number of agents have been removed from a subscription, this can have disastrous effects on round robin scheduling.

Because of the asynchronous nature of Sensu, it's difficult to detect and deal with this problem. Sensu doesn't associate round robin scheduling requests with check results in any way. So after sending a check request to a subscription, Sensu will never know if any result was delivered for it.

In the commercial distribution of Sensu, we have developed a new round robin scheduling mechanism on top of postgresql, which isn't supported by the OSS sensu-go. We expect this scheduling mechanism to be reliable across sensu-backend and postgres restarts, as there are no in-memory portions of the design.

There may be a way to solve the issue for sensu-go as well, by correlating incoming events with their scheduler, but this would involve a ground-up redesign of the sensu-go scheduler and we aren't ready to tackle that at this time.

Users that need to rely on sensu-go's round robin check scheduling can mitigate the problem somewhat by following the advice below:

- Use only high performance storage for etcd (NVMe SSDs).
- Run etcd separately from sensu-backend, in order to maximize its uptime.
- Run a single etcd node (do not cluster).
- Ensure that the steady-state CPU usage of etcd is low enough to handle bursts of traffic.
- Keep the number of agents per subscriptions relatively small.
- Don't use ephemeral agents (deregister: true). This can lead to more ring removals than with persistent agents (deregister: false), and destabilize round-robin scheduling.

**Comment by portertech** [comment]
At this time, we have a fix that requires the commercial postgres store. Commercial distribution users get consistent and reliable round-robin scheduling. As @echlebek mentioned in his summary, there may be a way to solve the issue in sensu-go, we welcome ideas and contributions. Closing this issue for now.
